WIP: Update route reconciler to create ClusterIngress instead of VirtualService

lichuqiang opened the pull request:
<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Part of #1963
Follow-up #2189

## Proposed Changes

  * helper funcs to construct clusteringress
  * create placeholder service according to ingress status
  * update route reconciler to create ClusterIngress instead of VirtualService

**Release Note**
<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->
```
update route reconciler to create ClusterIngress instead of VirtualService
```

It is developed on top of https://github.com/knative/serving/pull/2189, I'll have it rebased after https://github.com/knative/serving/pull/2189 merged

/area networking

lichuqiang commented on pkg/reconciler/v1alpha1/clusteringress/resources/virtual_service.go:
```diff
@@ -0,0 +1,153 @@
+limitations under the License.
+*/
+
+package resources
```
we do have package comment in doc.go file

tcnghia commented:
/retest

lichuqiang commented on pkg/reconciler/v1alpha1/route/route.go:
```diff
@@ -110,12 +114,10 @@ func NewController(
+
+	clusterIngressInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.enqueueOwnerRoute(impl),
+		UpdateFunc: controller.PassNew(c.enqueueOwnerRoute(impl)),
```
I'm surprised at the fact that the controllers (not only networking related controllers, but all of the existing controllers) do not watch deletion events from child resources. Any special reasons that force us to make it this way?

I think it fairly reasonable and necessary for a controller to watch the events, and create one if its resource got deleted mistakenly.

lichuqiang commented:
It is being blocked by a weird issue: ingress reconciler can not update ClusterIngress status to specify loadbalancer as expected.
The `DomainInternal` field is specified for sure in [local ingress instancc](https://github.com/knative/serving/blob/master/pkg/reconciler/v1alpha1/clusteringress/clusteringress.go#L139), but after update, we turn out get an empty item in its Ingress slice. As a result, we route controller can not create the placeholder service.
Not sure it's a CRD bug, or we use it in a wrongly somewhere. I've filed an issue in k8s community https://github.com/kubernetes/kubernetes/issues/70069.

The e2e test is successing, because [here](https://github.com/knative/serving/blob/master/pkg/reconciler/v1alpha1/clusteringress/clusteringress.go#L139) when updating status in ingress reconciler, we do not make a copy before modify the cached resource, thus, route reconciler has the chance to fetch loadbalancer info from the informer, and placeholder service got created.  I believe this should also be a bug that need fix: we need to make a copy before any modification on cache resources from shared informer, in case the cache get soiled.

As the updatestatus func is basicly copied from other controllers, we might need to inspect other controllers as well.

krancour commented on cmd/controller/main.go:
```diff
@@ -166,7 +166,7 @@ func main() {
 			configurationInformer,
 			revisionInformer,
 			coreServiceInformer,
-			virtualServiceInformer,
```
I notice in this same file that `virtualServiceInformer` is still defined and synced. Can't both of those goes away?

krancour replied:
Nevermind. It's still used by the cluster ingress controller.

krancour replied:
However... shouldn't the cluster ingress controller _not_ be started by this controller manager?

If the end-game is decoupling from Istio, then I assume operators deploying knative serving should choose their own cluster ingress controller in the same way they'd have chosen their own (plain old) ingress controller with plain k8s.

krancour replied:
Maybe that happens in a follow-up?

mattmoor replied:
@krancour That's a later increment.

lichuqiang replied:
good point. We'll have a special work to decide the shape of underlying network programming controllers after this.

lichuqiang commented:
The PR is ready for review @tcnghia @mattmoor @ZhiminXiang

mattmoor commented on pkg/reconciler/v1alpha1/route/resources/service.go:
```diff
@@ -38,9 +48,36 @@ func MakeK8sService(route *v1alpha1.Route) *corev1.Service {
+	if ingressStatus.LoadBalancer == nil || len(ingressStatus.LoadBalancer.Ingress) == 0 {
+		return nil, errLoadBalancerNotFound
+	}
+	// TODO(lichuqiang): what to do when we have multiple ingresses here?
```
Feels like something we should bias towards erroring on instead of silently ignoring?

mattmoor commented on pkg/reconciler/v1alpha1/route/resources/cluster_ingress.go:
```diff
@@ -0,0 +1,179 @@
+		ObjectMeta: metav1.ObjectMeta{
+			// As ClusterIngress resource is cluster-scoped,
+			// here we use GenerateName to avoid conflict.
+			GenerateName: fmt.Sprintf("%s-", names.ClusterIngress(r)),
```
See my comment in the name package, I think we should use `names.ClusterIngressPrefix` and include the hyphen there.

lichuqiang commented on pkg/apis/serving/v1alpha1/route_types.go:
```diff
@@ -107,9 +109,13 @@ const (
+
+	// RouteConditionIngressReady is set to False when the
+	// ClusterIngress fails to become Ready.
+	RouteConditionIngressReady duckv1alpha1.ConditionType = "IngressReady"
```
Ah, thanks for reminding.
So the doc is automatically updated (through tools/scripts), or updated manually?

mattmoor replied:
Manual.  If you know of good tooling, let us know :)

lichuqiang replied:
Still I did not update the doc. Because as it indicates [here](https://github.com/knative/serving/blob/master/pkg/reconciler/v1alpha1/clusteringress/clusteringress.go#L152), currently we can only see `Ready` and `Unknown` in ingress status, and no known errors yet. 
I'd like to revisit this when we have way to probe virtualservice status.

tcnghia replied:
Let me bring this up in WG meet up today.  Yes, I'd love to see more correctness in Route.Status.Ready.

lichuqiang commented on pkg/reconciler/v1alpha1/route/reconcile_resources_test.go:
```diff
@@ -20,78 +20,66 @@ import (
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "test-route",
 			Namespace: "test-ns",
-			Labels: map[string]string{
```
I did not find the place that in need of this.
And tried to remove it, and saw no side effect？

mattmoor replied:
I see, that sounds fine :)

lichuqiang commented on pkg/reconciler/v1alpha1/route/reconcile_resources.go:
```diff
@@ -19,63 +19,101 @@ package route
+		return nil, apierrs.NewNotFound(v1alpha1.Resource("clusteringress"), resourcenames.ClusterIngress(route))
+	}
+	// Sort the ingresses by name, and use the first one for consistency.
+	sort.Sort(ingressByName(ingresses))
```
I treat the unexpected scenarios of ingress resource (both here and some of the scenarios you mentioned below) relatively “loosely”, for I thought users can create things manually.

If we treat the ingress resource absolutely "internal", then I'm not opposed to treat these all as errors

lichuqiang replied:
k, you convinced me :)

lichuqiang commented on pkg/reconciler/v1alpha1/route/route_test.go:
```diff
@@ -242,30 +261,28 @@ func addResourcesToInformers(
+			t.Errorf("Services.Get(%v) = %v", ksName, err)
+		}
+		kubeInformer.Core().V1().Services().Informer().GetIndexer().Add(service)
+	*/
```
Ah, something I forgot to cleanup

lichuqiang commented on pkg/reconciler/v1alpha1/route/resources/names/names.go:
```diff
@@ -34,10 +34,11 @@ func K8sService(route *v1alpha1.Route) string {
 }
+
+// ClusterIngress returns the name of the ClusterIngress child resource for given Route.
+func ClusterIngress(route *v1alpha1.Route) string {
```
Ack

lichuqiang commented on pkg/reconciler/v1alpha1/route/resources/cluster_ingress.go:
```diff
@@ -0,0 +1,179 @@
+
+// MakeClusterIngress creates ClusterIngress to set up routing rules. Such ClusterIngress specifies
+// which Hosts that it applies to, as well as the routing rules.
+func MakeClusterIngress(r *servingv1alpha1.Route, tc *traffic.TrafficConfig) *v1alpha1.ClusterIngress {
```
Yeah, I tried to chunk things well into separate commits.
But separate PRs should be more friendly anyway.

lichuqiang commented on pkg/reconciler/v1alpha1/route/resources/cluster_ingress.go:
```diff
@@ -0,0 +1,179 @@
+		// Nameless traffic targets correspond to many domains: the
+		// Route.Status.Domain, and also various names of the Route's
+		// headless Service.
+		domains = []string{domain,
```
ha, I made it this way mainly because I don't expect to see dedup everywhere.
It's a minor issue on code style preference anyway.
Since you have strong persistence on code conciseness, Let's make it the way you suggested :)

lichuqiang replied:
Ah, I see what you mean.
The domains do not indeed need `dedup` for named target, so it's fine.

lichuqiang commented:
Hopefully I've addressed most of the comments above (except for the doc one).
And I've also open this separate PR for review convenience https://github.com/knative/serving/pull/2298

tcnghia commented:
@lichuqiang Can you please rebase?

lichuqiang commented:
/retest

lichuqiang commented:
@tcnghia Code rebased :)

mattmoor commented on pkg/reconciler/v1alpha1/route/reconcile_resources.go:
```diff
@@ -40,51 +41,89 @@ import (
 	ns := route.Namespace
 	name := resourcenames.K8sService(route)
 
+	ingress, err := c.getClusterIngressForRoute(route)
```
Would it be possible to pass this in from `reconcileClusterIngress`?

mattmoor commented:
@lichuqiang thanks a lot for driving this!
